instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,382 @@
+	log.Printf("[INFO] Updating the config of Kafka topic %s in %s.", topic, cluster_id)
+
+	// This is for when only replication_factor or partitions are changed, but we don't support changing them currently
+	if !d.HasChange("config") {
+		log.Printf("[INFO] Currently we only supprot updating topic's config. There are no changes in topic %s's config.", topic)
+		return nil
+	}
+	// This is for when there is no config{} block in the resource, or the whole config{} is removed from the resource.
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
+	if err != nil {
+		return err
+	}
+	if len(changedConfigMap) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		err = resourceKafkaTopicRead(d, meta)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var kafkaTopicConfigOptions KafkaTopicConfigOptions
+	err = mapstructure.Decode(changedConfigMap, &kafkaTopicConfigOptions)
+	if err != nil {
+		return err
+	}
+
+	updateKafkaTopicRequest := UpdateKafkaTopicRequest{
+		Config: &kafkaTopicConfigOptions,
+	}
+
+	jsonStr, err := json.Marshal(updateKafkaTopicRequest)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic update request: %s", err)
+	}
+
+	err = client.UpdateKafkaTopic(cluster_id, topic, jsonStr)
+	if err != nil {
+		return fmt.Errorf("[Error] Error updating the config of Kafka topic %s: %s", topic, err)
+	}
+
+	//Reading updated configs
+	err = resourceKafkaTopicRead(d, meta)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] The configs of Kafka topic %s has been updated.", topic)
+	return nil
+}
+
+func getChangedConfigMap(d *schema.ResourceData) (map[string]interface{}, error) {
+	newConfig := d.Get("config").([]interface{})[0].(map[string]interface{})
+	changedConfigMap := make(map[string]interface{}, 0)
+	var kafkaTopicConfig KafkaTopicConfigOptions
+	decodedConfigMap := make(map[string]interface{}, 0) // This map stores all non-zero values and all the pointers
+
+	// Decode twice here because zero-value non-pointers need to be removed
+	err := mapstructure.Decode(newConfig, &kafkaTopicConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = mapstructure.Decode(kafkaTopicConfig, &decodedConfigMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// The following block is to get what configs are indeed changed
+	for k, v := range newConfig {
+		key := fmt.Sprintf("config.0.%s", k)
+		valueOfV := reflect.ValueOf(decodedConfigMap[k])
+		if d.HasChange(key) {
+			before, after := d.GetChange(key)
+			isBeforeZeroValue := before == reflect.Zero(reflect.TypeOf(before)).Interface()
+			isAfterZeroValue := after == reflect.Zero(reflect.TypeOf(after)).Interface()
+			if !isBeforeZeroValue && !isAfterZeroValue {
+				changedConfigMap[k] = v
+			}
+			if !isBeforeZeroValue && isAfterZeroValue && (valueOfV.Kind() == reflect.Ptr) {
```

Comment:
what about keys that can be genuinely 0? like `flush.messages`?

Reply:
This is a really good question, I thought a lot about this, that's also the reason why I set some of the configs as * pointers. My thought was to set the configs that might be set as zero-value and configs that have a default zero-value to pointers. So I look at this https://kafka.apache.org/23/documentation.html#topicconfigs, but I found there are many of them. But for configs like flush.messages, which has a very huge default number, I assume it will not be set to 0.

Reply:
that makes sense to me